### ISSUE TITLE
QueryExpression Condition.In throwing errors when referencing EntityReference attribute

### ISSUE THREAD

**Issue by luifer** [jordimontana82/fake-xrm-easy#180]
Hi,

If you try to use the Condition.In on a QueryExpression with an EntityReference you get 2 different types of errors depending on how the proxy types where generated:

1) The standard generation throws the exception: 
System.InvalidOperationException : The binary operator Equal is not defined for the types 'System.Guid' and 'System.Guid[]'.

2) When generated using the developer extensions (https://msdn.microsoft.com/en-us/library/gg695784.aspx) it throws:
System.InvalidCastException : Unable to cast object of type 'Microsoft.Xrm.Sdk.EntityReference' to type 'System.Guid[]'

These errors are using CRM2015 - have not tested on other versions. I have attached a sample test project that shows the 2 issues. 
[FakeXrmEasyFailingTest.zip](https://github.com/jordimontana82/fake-xrm-easy/files/1002746/FakeXrmEasyFailingTest.zip)

Thanks
Luis

**Comment by jordimontana82** [comment]
Thanks Luis,

I'm just copying them here for easy review. I'll take a look.

    [Test]
        public void Test()
        {
            var fakedContext = new XrmFakedContext { ProxyTypesAssembly = typeof(Account).Assembly };
            var fakedService = fakedContext.GetOrganizationService();

            var account = new Account
            {
                Id = Guid.NewGuid(),
                OriginatingLeadId = new EntityReference("lead", Guid.NewGuid())
            };

            fakedContext.Initialize(new List<Entity> { account });

            var ids = new[] { account.OriginatingLeadId.Id };
            
            var qe = new QueryExpression(Account.EntityLogicalName);
            qe.Criteria.AddCondition("originatingleadid", ConditionOperator.In, ids);

            var entities = fakedService.RetrieveMultiple(qe).Entities;

            Assert.That(entities.Count, Is.EqualTo(1));
        }

     [Test]
        public void Test()
        {
            var fakedContext = new XrmFakedContext { ProxyTypesAssembly = typeof(Account).Assembly };
            var fakedService = fakedContext.GetOrganizationService();

            var account = new Account
            {
                Id = Guid.NewGuid(),
                OriginatingLeadId = new EntityReference("lead", Guid.NewGuid())
            };

            fakedContext.Initialize(new List<Entity> { account });

            var ids = new[] { account.OriginatingLeadId.Id };
            
            var qe = new QueryExpression(Account.EntityLogicalName);
            qe.Criteria.AddCondition("originatingleadid", ConditionOperator.In, ids);

            var entities = fakedService.RetrieveMultiple(qe).Entities;

            Assert.That(entities.Count, Is.EqualTo(1));
        }

**Comment by eNeRGy164** [comment]
Hi @luifer, can you verify if this is fixed now?

**Comment by jordimontana82** [comment]
Thx for this fix @eNeRGy164 . @luifer , if you are happy then we'll close the issue.

**Comment by luifer** [comment]
Hi, @jordimontana82 @eNeRGy164 

Thanks for looking into it.

The latest version fixes just one of my tests, the one that uses the standard proxy types used by svcutils.

The remaining failing test is throwing an invalid cast exception. I have debugged the issue and it looks like the problem is in the class XrmFakedContext and method 

`GetAppropiateCastExpressionBasedOnAttributeTypeOrValue(Expression input, object value, Type attributeType)`

This method contains the following check:
```
if (attributeType == typeof(EntityReference))
return GetAppropiateCastExpressionBasedOnEntityReference(input, value);
```

the attribute type when running the test is CrmEntityReference so it skips the if and just executes:
return GetAppropiateCastExpressionDefault(input, value); 

if I change the if statement to:
`if (attributeType == typeof(EntityReference) || attributeType == typeof(CrmEntityReference) )`

then both test pass. Not sure is that is the correct way of fixing this issue as I have also noticed the following at the beginning of the method:
```
#if FAKE_XRM_EASY
                    if (attributeType == typeof(Microsoft.Xrm.Client.CrmEntityReference))
                            return GetAppropiateCastExpressionBasedGuid(input);
#endif 
```
I am using the project for FakeXrmEasy.2015 as I am running this code as an exe against a CRM 2015 instance. 

Should I be using FaxeXrmEasy with no version number? I have noticed that the code does the check for #if FAKE_XRM_EASY in a few places to add support for CrmEntityRefence. Doing that throws an error when trying to load the sdk assemblies as it expects a higher version than then ones I reference (2015).

The following change also fixes the issue:
```
#if FAKE_XRM_EASY || FAKE_XRM_EASY_2013 || FAKE_XRM_EASY_2015
                if (attributeType == typeof(Microsoft.Xrm.Client.CrmEntityReference))
                            return GetAppropiateCastExpressionBasedGuid(input);
#endif
```
I have not included 2016 or 365 as from what I understand the CrmEntity and CrmEntityReference classes have been deprecated in CRM 2016 and 365 as the proxy types will just use the standard Entity and EntityReferences. The ms documentation seems to support that since there is no option for these here:
https://msdn.microsoft.com/en-us/library/microsoft.xrm.client.crmentityreference(v=crm.7).aspx

Sorry for the long post. Hope you find it useful in fixing the code. For my immediate needs any of these 2 fixes will do but you may want to do a deeper review into the usage of CrmEntityReference.

Thanks
Luis

**Comment by jordimontana82** [comment]
Thanks Luis,

Indeed, Xrm.Client is deprecated in favor of the XrmTolling assemblies, you're right.

We could possibly include the fix you mention with a conditional compilation statement for versions <= 2015. 

    if (attributeType == typeof(EntityReference) || attributeType == typeof(CrmEntityReference) )


Like:

    #if FAKE_XRM_EASY ||FAKE_XRM_EASY_2013 || FAKE_XRM_EASY_2015
       if (attributeType == typeof(EntityReference) || attributeType == typeof(CrmEntityReference) )
    #else 
       if (attributeType == typeof(EntityReference) )
    #endif 
So yeah, you shouldn't have to use FakeXrmEasy without the version, if you are using CRM 2015 then FakeXrmEasy.2015 is your package. These versions are aiming at maintaining backward compatible versions so that we could reuse as much code as possible, while still allowing for newere messages  in the newer packages.

What do you think guys?

**Comment by eNeRGy164** [comment]
I think the conditional compilation with the mentioned versions is the way to go.

**Comment by luifer** [comment]
Thinking about it, not sure why you need the different versions. You could
just always check if it's either an entityreferebce or a
CrmEntityReference. The behaviour doesn't really change across versions
just that the type is deprecated plus you may find the odd case of someone
using a previously generated proxy types with the new sdk although not sure
if that is possible.

Luis

**Comment by jordimontana82** [comment]
Basically because latest versions don't reference Microsoft.Xrm.Client.dll anymore and so it wouldn't even compile.

**Comment by eNeRGy164** [comment]
Is this still an issue? Because I can't recreate this scenario with the current version.

**Comment by luifer** [comment]
Yes it is still an issue.

I have attached the updated sample code with unit tests using the latest version of fake-xrm-easy. The proposed solution discussed would fix the test failing test.

The 2 tests are written in nunit rather than mstest but they would be trivial to convert. Please let me know if you need any help with that.

Thanks for looking into it.

[FakeXrmEasyFailingTest.zip](https://github.com/jordimontana82/fake-xrm-easy/files/1223720/FakeXrmEasyFailingTest.zip)

**Comment by eNeRGy164** [comment]
The strange thing is that it indeed fails when I open your solution, when I copy it inside the FakeXrmEasy Test project it stops failing...

**Comment by luifer** [comment]
It could be that when you copy the tests you stop using the proxy types as
generated on my test solution. It only fails if you generate the proxy with
the client extensions as mentioned in my first post.

On 15 Aug 2017 7:47 am, "Michaël Hompus" <notifications@github.com> wrote:

The strange thing is that it indeed fails when I open your solution, when I
copy it inside the FakeXrmEasy Test project it stops failing...

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/jordimontana82/fake-xrm-easy/issues/180#issuecomment-322390097>,
or mute the thread

**Comment by n-o-r-m-a-n** [comment]
I encounter the same or at least a simular issue when filtering any LookupValue.

Context: CRM 2015, early bound entities (XrmToolBox, DLaB.Xrm.EarlyBoundGenerator.1.2017.3.23).
Query: Quote with condition on CustomerId equals <guid>
Result: InvalidCastException

possible fix: adding FAKE_XRM_EASY_2015 in XrmFakedContext.GetAppropiateCastExpressionBasedOnAttributeTypeOrValue():

```
protected static Expression GetAppropiateCastExpressionBasedOnAttributeTypeOrValue(Expression input, object value, Type attributeType)
        {
            if (attributeType != null)
            {

#if FAKE_XRM_EASY **|| FAKE_XRM_EASY_2015**
                    if (attributeType == typeof(Microsoft.Xrm.Client.CrmEntityReference))
                            return GetAppropiateCastExpressionBasedGuid(input);
#endif
                if (attributeType == typeof(Guid))
                    return GetAppropiateCastExpressionBasedGuid(input);
```

=> as CrmEntityReference is present in 2015, why is this switch not present here?

**Comment by jordimontana82** [comment]
Hi @n-o-r-m-a-n 

I suppose we could extend it for 2013 and 2015 yep, at least based on [this link](https://msdn.microsoft.com/en-us/library/microsoft.xrm.client(v=crm.7).aspx). Although Xrm.Client is deprecated now in favour of XrmTooling.

**Comment by jordimontana82** [comment]
Fixed in PR #264
